kernel/src/checkpoint/log_replay.rs

```diff
@@ -0,0 +1,603 @@
+//!
+//! - [`CheckpointLogReplayProcessor`]: Implements the [`LogReplayProcessor`] trait and orchestrates
+//!   the overall process. For each batch of log actions, it:
+//!   1. Creates a visitor with the current deduplication state
+//!   2. Applies the visitor to filter actions in the batch
+//!   3. Updates counters and state for cross-batch deduplication
+//!   4. Produces a [`CheckpointData`] result which includes a selection vector indicating which
+//!      actions should be included in the checkpoint file
+use std::collections::HashSet;
+use std::sync::LazyLock;
+
+use crate::engine_data::{GetData, RowVisitor, TypedGetData as _};
+use crate::log_replay::{FileActionDeduplicator, FileActionKey};
+use crate::schema::{column_name, ColumnName, ColumnNamesAndTypes, DataType};
+use crate::utils::require;
+use crate::{DeltaResult, Error};
+
+/// A visitor that filters actions for inclusion in a V1 spec checkpoint file.
+///
+/// This visitor processes actions in newest-to-oldest order (as they appear in log
+/// replay) and applies deduplication logic for both file and non-file actions to
+/// produce the minimal state representation for the table.
+///
+/// # File Action Filtering
+/// - Keeps only the first occurrence of each unique (path, dvId) pair
+/// - Excludes expired tombstone remove actions (where deletionTimestamp ≤ minimumFileRetentionTimestamp)
+/// - Add actions represent files present in the table
+/// - Unexpired remove actions represent tombstones still needed for consistency
+///
+/// # Non-File Action Filtering
+/// - Keeps only the first protocol action (newest version)
+/// - Keeps only the first metadata action (most recent table metadata)
+/// - Keeps only the first transaction action for each unique app ID
+///
+/// # Excluded Actions
+/// CommitInfo, CDC, and CheckpointMetadata actions should not appear in the action
+/// batches processed by this visitor, as they are excluded by the schema used to
+/// read the log files upstream. If present, they will be ignored by the visitor.
+/// Sidecar actions should also be excluded—when encountered in the log, the
+/// corresponding sidecar files are read to extract the referenced file actions,
+/// which are then included directly in the action stream instead of the sidecar actions themselves.
+///
+/// The resulting filtered set of actions represents the minimal set needed to reconstruct
+/// the latest valid state of the table at the checkpointed version.
+pub(crate) struct CheckpointVisitor<'seen> {
+    // Deduplicates file actions
+    deduplicator: FileActionDeduplicator<'seen>,
+    // Tracks which rows to include in the final output
+    selection_vector: Vec<bool>,
+    // TODO: _last_checkpoint schema should be updated to use u64 instead of i64
+    // for fields that are not expected to be negative. (Issue #786)
+    // i64 to match the `_last_checkpoint` file schema
+    total_file_actions: i64,
```

Comment:
nit: consider renaming (and below) totals to `count`
```suggestion
    file_actions_count: i64,
```